### ISSUE TITLE
MIR-borrowck: (unsound) fails to see error in borrowck-init-in-fru.rs

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#44833]
Many MIR-borrowck bugs have been categorized as either extra errors or poor diagnostics, but this is a case where MIR borrowck is simply failing to flag *the* main error in the code in question.

We see below that AST-borrowck is flagging the error while MIR-borrowck says nothing. (This is *atop* the mir-borrowck4 branch from https://github.com/rust-lang/rust/pull/44736 )

```
% ./objdir-dbgopt/build/x86_64-unknown-linux-gnu/stage1/bin/rustc -Z emit-end-regions \
    -Z borrowck-mir ./src/test/compile-fail/borrowck/borrowck-init-in-fru.rs
error[E0381]: use of possibly uninitialized variable: `origin.y` (Ast)
  --> ./src/test/compile-fail/borrowck/borrowck-init-in-fru.rs:19:30
   |
19 |     origin = point {x: 10,.. origin}; //~ ERROR use of possibly uninitialized variable: `origin.y`
   |                              ^^^^^^ use of possibly uninitialized `origin.y`

error: aborting due to previous error
```

**Comment by mikhail-m1** [comment]
working on it

**Comment by mikhail-m1** [comment]
simplified 
```
struct Foo(isize);

fn main() {
    let mut a: Foo;
    let _ = Foo(a.1);
}
```

**Comment by pnkfelix** [comment]
(I think I'm going to take the lead on this bug at this point; PR #45025 doesn't have the *exact* right solution yet, but I think the work there is part of the answer.)
